### DISCUSSION TITLE
A Gift from Knowledge Distillation

### DISCUSSION THREAD

**Discussion by lansiyuan3** [yoshitomo-matsubara/torchdistill#366] [Q&A]
HI @yoshitomo-matsubara ：
 I'm looking for the code related to the paper A Gift from Knowledge Distillation, but I don't clear the location of the paper in your project. Or how you categorize these algorithms in general

**Comment by yoshitomo-matsubara** [comment]
Hi @lansiyuan3 

With the current PyPI package (torchdistill==0.3.3), you can run it using [this config file](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/legacy/sample/ilsvrc2012/multi_stage/fsp/resnet18_from_resnet34.yaml) using [this script ](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/examples/legacy/image_classification.py)

Note that you'll need to tune the hyperparameters as it's just a sample config (as explained [here](https://github.com/yoshitomo-matsubara/torchdistill/tree/main/configs#sample)) because the original paper did not test the method for ImageNet dataset

**Reply by yoshitomo-matsubara** [reply]
Hi @lansiyuan3 

Please use Reply for follow-up questions  your "Comment" is posted as "answer"

> I want to insert the attention mechanism in the neural network, so do I need to modify the corresponding network structure in yaml?
Or should the function insert the network into the attention mechanism？

If you use the attention mechanism as part of model, I'd define a new student model class and register a function to instantiate the new class like [this](https://github.com/yoshitomo-matsubara/torchdistill/discussions/278#discussioncomment-4700064).

**Reply by lansiyuan3** [reply]
Hi@yoshitomo-matsubara:
now ,i want to build own dataset for image_classification.but i have two question.
one:
if i craete a function for own dataset, the Returned objects not suitable for your dataloader
two:
i want to coverting my dataset to cifar10 format,then load it 
could you give me some advice?

**Reply by yoshitomo-matsubara** [reply]
Hi @lansiyuan3 

Share your function to return your own dataset and explain how it's different from dataset classes in torchvision (say ImageFolder and CIFAR10)

Otherwise, I cannot understand what you want to do with torchdistill

**Reply by lansiyuan3** [reply]
I haven't written the function yet, I want to do an image classification with my dataset, but you did it with cifar10.
I'm not sure how to load it into your DistillationBox with my dataset

**Reply by yoshitomo-matsubara** [reply]
In your case (image classification with torchdistill v0.3.3), if you cannot reuse [any image classification dataset classes in torchvision](https://pytorch.org/vision/stable/datasets.html) , you define your dataset class with a dataset registry function in some random python file (e.g., custom.py) like this

```python
from torchvision.datasets import VisionDataset
from torchdistill.datasets.registry import register_dataset


@register_dataset
class MyDataset(VisionDataset): # It doesn't have to be MyDataset, basically you can use any class name you like
...

```

Also, you should make sure that your dataset is compatible with other image classification datasets e.g., data loader should return a pair of image tensor batch and target label batch.

Once you confirm that, you have `examples/legacy/image_classification.py` import custom.py i.e., `import custom` and replace dataset entries in your yaml file (e.g., [here](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/legacy/sample/cifar10/kd/resnet20_from_densenet_bc_k12_depth100-final_run.yaml#L1-L44)) with `MyDataset` as type

**Reply by yoshitomo-matsubara** [reply]
Again, please keep the discussion in this thread and do not start a new thread (it will be treated as an answer candidate). This is the 2nd time

From the information, I cannot understand what caused the error

Share more details (which script did you edit and run? what changes did you make?) and code + yaml config you used

**Reply by lansiyuan3** [reply]
Sorry forgot your previous reminder, I created my own dataset in your image categorization example under the distillation framework and he only has 8 categories. I used Resnet18 from Resnet34 ,he produced this error and would like to have your advice!

![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/51487857/bf39e378-ef4e-4ab7-a20c-2037113b7c7e)

**Reply by yoshitomo-matsubara** [reply]
The error message is almost the same as the previous one you deleted. If you cannot share your code or yaml config, I have to guess that you're using `pretrained: True` for models (which should `num_classes: 1000`), but you put `num_classes: 8`, which is not the error from torchdistill. In that case, you are trying to use a teacher model trained on ImageNet for a student to be trained on your own dataset, which sounds weird to me.

**Reply by lansiyuan3** [reply]
Right, I started training with the pre-trained model and created my own training set, but my training set only has 8 categories
![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/51487857/d5b39ec1-aba7-463d-a5ca-63cd7e51d1bf)
 I want to modify my pre-training set like below
![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/51487857/91898b65-fd6e-4694-8f45-18fdcb2d8648)
I modified the yaml file and found that he didn't work.
I'd like your advice.

**Reply by yoshitomo-matsubara** [reply]
Notice that the configuration implies that your teacher model's `fc` will be randomly initialized and student model cannot learn from the random teacher. Thus, you will need to train a teacher model with `num_classes=8` on your own dataset at first (before knowledge distillation experiment).

Also, it seems that the yaml config (e.g., `kwargs`, `src_ckpt`, `dst_ckpt`, `weights` and more) is not compatible with torchdistill v0.3.3. You should start with legacy config files as I suggested in [this thread](https://github.com/yoshitomo-matsubara/torchdistill/discussions/366#discussioncomment-6392293) you do so.

Until you finish these two steps, I guess I cannot help you

**Reply by lansiyuan3** [reply]
Thank you for your suggestion, after reading your suggestion, I'm going to work with the following idea: I'm going to use your offcial.py (as shown below) and then load my own pre-trained ResNet34, save it as ResNet34.pt, and then proceed with the knowledge distillation. Not sure if this is the right way of thinking
![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/51487857/364a1364-40cd-4f6e-b6a2-ba75ca11a127)

**Reply by yoshitomo-matsubara** [reply]
You can pass your model weights for teacher as its `ckpt: your_file_path_to_model_weight` in the yaml config for the legacy script I repeatedly suggested.

Again, the screenshot still shows non-legacy code, which I cannot help you resolve your problem right now. I do strongly suggest that you use the PyPI version of torchdistill (pip3 install torchdistill) instead of copying my code as a new repository and directly editing code under `torchdistill/`. (I saw your repo.)

**Reply by lansiyuan3** [reply]
HI @yoshitomo-matsubara ：
  I now have a new problem in image_clssification. I am using MultiStagesDistillationBox and the number of training rounds in the first stage is 10 and the number of training rounds in the second stage is 5. But when I train to epoch=9 in the first stage, I get this error message [setting requires_grad=True on inference tensor outside InferenceMode is not allowed] , I debugged the program and found that this error occurs at in these places
![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/51487857/e2fc34b9-f57f-4a62-9945-5e2da5d2ae45)
![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/51487857/d38ce664-6559-4736-aefe-2c865e2e8f85)
![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/51487857/76c767a6-9f9f-41d2-aa1f-5a44ebe3e4d1)
I'm guessing it might be because the model is set to inference mode when it enters stage 2, and I'm hoping you can advise.

**Reply by lansiyuan3** [reply]
and about it
![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/51487857/cfbfef5c-68c3-4640-9d91-c569acb8c3f1)

**Reply by yoshitomo-matsubara** [reply]
Please read my previous comment carefully.

> Again, the screenshot still shows non-legacy code, which I cannot help you resolve your problem right now. I do strongly suggest that you use the PyPI version of torchdistill (pip3 install torchdistill) instead of copying my code as a new repository and directly editing code under torchdistill/. (I saw your repo.)

[The code you copied from the torchdistill repo](https://github.com/lansiyuan3/My-Pro-Dis/tree/main/torchdistill) is not ready to be used at this point, and currently I cannot give you advices if you are not using legacy code with torchdistill v0.3.3

**Reply by lansiyuan3** [reply]
I took your advice，using legacy code with torchdistill v0.3.3 at torchdistill/examples/legacy/image_classification.py,
But the training_box don't have post_process() in your legacy code.So I modified it to training_box.post_epoch_process()，and out of line with the aforementioned problems.

**Reply by yoshitomo-matsubara** [reply]
No, you didn't.
With v0.3.3, there should be [`post_process` in the example scripts](https://github.com/yoshitomo-matsubara/torchdistill/blob/v0.3.3/examples/image_classification.py#L132) instead of `post_epoch_process` 
[The legacy scripts with the latest version](https://github.com/yoshitomo-matsubara/torchdistill/tree/main/examples/legacy#L132) also use `post_process`

As I pointed out with a link above, you are copying the entire code from this repository and making changes.

**Reply by lansiyuan3** [reply]
HI @yoshitomo-matsubara ：
I think I'm taking your advice now.But I'm having this problem while training, and I realized that I'm having the same problem with cifar10, so I can rule out that it's not my dataloader!
![image](https://github.com/yoshitomo-matsubara/torchdistill/assets/51487857/677d2118-f102-4b08-b73e-7a2fac6bf630)

I'd like your advice.

**Reply by yoshitomo-matsubara** [reply]
1. Please do not use screenshots to show the logs or files, but use texts for better search experience to help others find this discussion when they face similar issues.
2. Provide more details. Without knowing what changes you made in scripts and yaml configs, I cannot help you. You're still copying all the code from this repository and I don't even know how you ran the script, which torchdistill v0.3.3 you used, locally installed torchdistill or PyPI package (i.e., pip3 install torchdistill)

**Reply by lansiyuan3** [reply]
First of all, I used this[config file](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/ilsvrc2012/fsp/resnet18_from_resnet34.yaml) and [this script](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/examples/torchvision/image_classification.py). All the experiments are based on[this](https://github.com/yoshitomo-matsubara/torchdistill)
And then out of this error:
Files already downloaded and verified
2023/08/03 10:17:48	INFO	torchdistill.common.main_util	Getting `ResNet34_Weights` from `torchvision.models.resnet`
2023/08/03 10:17:48	INFO	torchdistill.common.main_util	Getting `SequentialSampler` from `torch.utils.data`
2023/08/03 10:17:48	INFO	torchdistill.common.main_util	Getting `RandomSampler` from `torch.utils.data`
2023/08/03 10:17:48	INFO	torchdistill.common.main_util	ckpt file path is None
2023/08/03 10:17:48	INFO	torchdistill.common.main_util	ckpt file path is None
2023/08/03 10:17:48	INFO	__main__	Start training
2023/08/03 10:17:50	INFO	torchdistill.core.distillation	[teacher model]
2023/08/03 10:17:50	INFO	torchdistill.models.util	Redesigning the teacher model with ['conv1', 'bn1', 'relu', 'maxpool', 'layer1', 'layer2', 'layer3', 'layer4']
2023/08/03 10:17:50	INFO	torchdistill.core.distillation	[student model]
2023/08/03 10:17:50	INFO	torchdistill.models.util	Redesigning the student model with ['conv1', 'bn1', 'relu', 'maxpool', 'layer1', 'layer2', 'layer3', 'layer4']
2023/08/03 10:17:50	INFO	torchdistill.core.distillation	Loss = 1.0 * FSPLoss()
2023/08/03 10:17:50	INFO	torchdistill.core.distillation	Freezing the whole teacher model
2023/08/03 10:17:50	INFO	torchdistill.core.distillation	Started stage 1
Traceback (most recent call last):
  File "D:\Users\Miclan\PycharmProjects\pythonProject2\torchdistill-main\torchdistill-main\examples\torchvision\image_classification.py", line 213, in <module>
    main(argparser.parse_args())
  File "D:\Users\Miclan\PycharmProjects\pythonProject2\torchdistill-main\torchdistill-main\examples\torchvision\image_classification.py", line 191, in main
    train(teacher_model, student_model, dataset_dict, dst_ckpt_file_path,
  File "D:\Users\Miclan\PycharmProjects\pythonProject2\torchdistill-main\torchdistill-main\examples\torchvision\image_classification.py", line 144, in train
    train_one_epoch(training_box, device, epoch, log_freq)
  File "D:\Users\Miclan\PycharmProjects\pythonProject2\torchdistill-main\torchdistill-main\examples\torchvision\image_classification.py", line 65, in train_one_epoch
    for sample_batch, targets, supp_dict in \
ValueError: not enough values to unpack (expected 3, got 2)
I don't know if this supp_dict returns any arguments，But it looks like **loss = training_box.forward_process(sample_batch, targets, supp_dict)** is needed

**Reply by yoshitomo-matsubara** [reply]
Again, you are not using legacy scripts (examples/legacy/) with PyPI version of torchdistill v0.3.3.
Do `pip3 install torchdistill`, confirm `torchdistill.__version__` returns `0.3.3` and use `examples/legacy/image_classification.py`.

training_box.forward_process is non-legacy implementation, which is not ready to be used right now

I've already emphasized this point at https://github.com/yoshitomo-matsubara/torchdistill/tree/main#important-notice , and you repeatedly ignore the message

**Reply by lansiyuan3** [reply]
I have made some modifications according to your request, and I think this should be correct. about [this script](https://github.com/lansiyuan3/Modulation/blob/main/torchdistill-main/torchdistill-main/examples/legacy/image_classification.py) and [yaml](https://github.com/lansiyuan3/Modulation/blob/main/torchdistill-main/torchdistill-main/configs/legacy/sample/cifar10/kd/resnet20_from_densenet_bc_k12_depth100-final_run.yaml)
I just want to experiment with whether this system can run and then add my own data loader, but I have found the following issues
`C:\Users\Miclan\anaconda3\python.exe D:\torchdistill-0.3.3\torchdistill-main\torchdistill-main\examples\legacy\image_classification.py 
2023/08/03 16:07:42	INFO	torchdistill.common.main_util	Not using distributed mode
2023/08/03 16:07:42	INFO	__main__	Namespace(adjust_lr=False, config='D:\\torchdistill-0.3.3\\torchdistill-main\\torchdistill-main\\configs\\legacy\\sample\\cifar10\\kd\\resnet20_from_densenet_bc_k12_depth100-final_run.yaml', device='cuda', dist_url='env://', log=None, log_config=False, seed=None, start_epoch=0, student_only=False, test_only=False, world_size=1)
2023/08/03 16:07:42	INFO	torchdistill.datasets.util	Loading train data
Downloading https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz to ./resource/dataset/cifar10\cifar-10-python.tar.gz
100%|██████████| 170498071/170498071 [00:50<00:00, 3351126.91it/s]
Extracting ./resource/dataset/cifar10\cifar-10-python.tar.gz to ./resource/dataset/cifar10
2023/08/03 16:08:40	INFO	torchdistill.datasets.util	dataset_id `cifar10/train`: 58.05756139755249 sec
2023/08/03 16:08:40	INFO	torchdistill.datasets.util	Loading val data
Files already downloaded and verified
2023/08/03 16:08:41	INFO	torchdistill.datasets.util	dataset_id `cifar10/val`: 0.5656769275665283 sec
2023/08/03 16:08:41	INFO	torchdistill.datasets.util	Loading test data
Files already downloaded and verified
2023/08/03 16:08:41	INFO	torchdistill.datasets.util	dataset_id `cifar10/test`: 0.5547194480895996 sec
2023/08/03 16:08:41	INFO	torchdistill.common.main_util	ckpt file is not found at `./resource/ckpt/cifar10/teacher/cifar10-densenet_bc_k12_depth100.pt`
2023/08/03 16:08:42	INFO	torchdistill.common.main_util	ckpt file is not found at `./resource/ckpt/cifar10/kd/cifar10-resnet20_from_densenet_bc_k12_depth100-final_run.pt`
2023/08/03 16:08:42	INFO	__main__	Start training
2023/08/03 16:08:42	INFO	torchdistill.models.util	[teacher model]
2023/08/03 16:08:42	INFO	torchdistill.models.util	Using the original teacher model
2023/08/03 16:08:42	INFO	torchdistill.models.util	[student model]
2023/08/03 16:08:42	INFO	torchdistill.models.util	Using the original student model
2023/08/03 16:08:42	INFO	torchdistill.core.distillation	Loss = 1.0 * OrgLoss
2023/08/03 16:08:42	INFO	torchdistill.core.distillation	Freezing the whole teacher model
Traceback (most recent call last):
  File "D:\torchdistill-0.3.3\torchdistill-main\torchdistill-main\examples\legacy\image_classification.py", line 188, in <module>
    main(argparser.parse_args())
  File "D:\torchdistill-0.3.3\torchdistill-main\torchdistill-main\examples\legacy\image_classification.py", line 169, in main
    train(teacher_model, student_model, dataset_dict, ckpt_file_path, device, device_ids, distributed, config, args)
  File "D:\torchdistill-0.3.3\torchdistill-main\torchdistill-main\examples\legacy\image_classification.py", line 125, in train
    train_one_epoch(training_box, device, epoch, log_freq)
  File "D:\torchdistill-0.3.3\torchdistill-main\torchdistill-main\examples\legacy\image_classification.py", line 62, in train_one_epoch
    for sample_batch, targets, supp_dict in \
ValueError: not enough values to unpack (expected 3, got 2)`
I discovered this issue while using your yaml file for training, which occurred in [here](https://github.com/lansiyuan3/Modulation/blob/main/torchdistill-main/torchdistill-main/examples/legacy/image_classification.py#L60-L65)
supp_dict cannot accept any parameters, but it is needed below

**Reply by yoshitomo-matsubara** [reply]
No, I think you still use the code you copied from this repository and made some modifications.

Following 
```shell
pip3 install torch==1.12.1 torchvision==0.13.1
pip3 install torchdistill
git clone https://github.com/yoshitomo-matsubara/torchdistill.git
```
I confirmed that the legacy script works for both training with/without teacher model with CIFAR-10 dataset
## Training with teacher
`python3 torchdistill/examples/legacy/image_classification.py --config torchdistill/configs/legacy/sample/cifar10/kd/resnet20_from_densenet_bc_k12_depth100-final_run.yaml --log log/cifar10/kd/resnet20_from_densenet_bc_k12_depth100-final_run.log`

```
2023/08/03 15:46:14	INFO	torchdistill.common.main_util	Not using distributed mode
2023/08/03 15:46:14	INFO	__main__	Namespace(config='torchdistill/configs/legacy/sample/cifar10/kd/resnet20_from_densenet_bc_k12_depth100-final_run.yaml', device='cuda', log='log/cifar10/kd/resnet20_from_densenet_bc_k12_depth100-final_run.log', start_epoch=0, seed=None, test_only=False, student_only=False, log_config=False, world_size=1, dist_url='env://', adjust_lr=False)
2023/08/03 15:46:14	INFO	torchdistill.datasets.util	Loading train data
Downloading https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz to ./resource/dataset/cifar10/cifar-10-python.tar.gz
100% 170498071/170498071 [00:03<00:00, 48270607.07it/s]
Extracting ./resource/dataset/cifar10/cifar-10-python.tar.gz to ./resource/dataset/cifar10
2023/08/03 15:46:20	INFO	torchdistill.datasets.util	dataset_id `cifar10/train`: 6.466401100158691 sec
2023/08/03 15:46:20	INFO	torchdistill.datasets.util	Loading val data
Files already downloaded and verified
2023/08/03 15:46:21	INFO	torchdistill.datasets.util	dataset_id `cifar10/val`: 0.6924517154693604 sec
2023/08/03 15:46:21	INFO	torchdistill.datasets.util	Loading test data
Files already downloaded and verified
2023/08/03 15:46:21	INFO	torchdistill.datasets.util	dataset_id `cifar10/test`: 0.71073317527771 sec
Downloading: "https://github.com/yoshitomo-matsubara/torchdistill/releases/download/v0.1.1/cifar10-densenet_bc_k12_depth100.pt" to /root/.cache/torch/hub/checkpoints/cifar10-densenet_bc_k12_depth100.pt
100% 3.23M/3.23M [00:00<00:00, 85.1MB/s]
2023/08/03 15:46:22	INFO	torchdistill.common.main_util	ckpt file is not found at `./resource/ckpt/cifar10/teacher/cifar10-densenet_bc_k12_depth100.pt`
2023/08/03 15:46:25	INFO	torchdistill.common.main_util	ckpt file is not found at `./resource/ckpt/cifar10/kd/cifar10-resnet20_from_densenet_bc_k12_depth100-final_run.pt`
2023/08/03 15:46:25	INFO	__main__	Start training
/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:563: UserWarning: This DataLoader will create 16 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(
2023/08/03 15:46:25	INFO	torchdistill.models.util	[teacher model]
2023/08/03 15:46:25	INFO	torchdistill.models.util	Using the original teacher model
2023/08/03 15:46:25	INFO	torchdistill.models.util	[student model]
2023/08/03 15:46:25	INFO	torchdistill.models.util	Using the original student model
2023/08/03 15:46:25	INFO	torchdistill.core.distillation	Loss = 1.0 * OrgLoss
2023/08/03 15:46:25	INFO	torchdistill.core.distillation	Freezing the whole teacher model
2023/08/03 15:46:29	INFO	torchdistill.misc.log	Epoch: [0]  [  0/391]  eta: 0:28:49  lr: 0.1  img/s: 67.25317249488089  loss: 4.7402 (4.7402)  time: 4.4227  data: 2.5192  max mem: 700
2023/08/03 15:46:40	INFO	torchdistill.misc.log	Epoch: [0]  [100/391]  eta: 0:00:44  lr: 0.1  img/s: 1200.8439511673578  loss: 3.5644 (3.9403)  time: 0.1067  data: 0.0002  max mem: 700
2023/08/03 15:46:51	INFO	torchdistill.misc.log	Epoch: [0]  [200/391]  eta: 0:00:25  lr: 0.1  img/s: 1198.011563479755  loss: 3.4226 (3.7161)  time: 0.1069  data: 0.0002  max mem: 700
2023/08/03 15:47:03	INFO	torchdistill.misc.log	Epoch: [0]  [300/391]  eta: 0:00:11  lr: 0.1  img/s: 1117.7498719587604  loss: 3.1425 (3.5530)  time: 0.1173  data: 0.0036  max mem: 700
2023/08/03 15:47:13	INFO	torchdistill.misc.log	Epoch: [0] Total time: 0:00:48
2023/08/03 15:47:14	INFO	torchdistill.misc.log	Validation:  [ 0/79]  eta: 0:01:23  acc1: 42.1875 (42.1875)  acc5: 89.8438 (89.8438)  time: 1.0631  data: 1.0392  max mem: 700
2023/08/03 15:47:17	INFO	torchdistill.misc.log	Validation: Total time: 0:00:04
2023/08/03 15:47:17	INFO	__main__	 * Acc@1 47.4200	Acc@5 92.9200

2023/08/03 15:47:17	INFO	__main__	Best top-1 accuracy: 0.0000 -> 47.4200
2023/08/03 15:47:17	INFO	__main__	Updating ckpt at ./resource/ckpt/cifar10/kd/cifar10-resnet20_from_densenet_bc_k12_depth100-final_run.pt
2023/08/03 15:47:20	INFO	torchdistill.misc.log	Epoch: [1]  [  0/391]  eta: 0:16:32  lr: 0.1  img/s: 197.798169355944  loss: 2.8337 (2.8337)  time: 2.5374  data: 1.8900  max mem: 700
2023/08/03 15:47:31	INFO	torchdistill.misc.log	Epoch: [1]  [100/391]  eta: 0:00:39  lr: 0.1  img/s: 1180.7846782192917  loss: 2.5681 (2.7318)  time: 0.1092  data: 0.0002  max mem: 700

```

## Training without teacher
`python3 torchdistill/examples/legacy/image_classification.py --config torchdistill/configs/legacy/sample/cifar10/ce/resnet20-final_run.yaml --log log/cifar10/ce/resnet20-final_run.log`

```
2023/08/03 15:38:45	INFO	torchdistill.common.main_util	Not using distributed mode
2023/08/03 15:38:45	INFO	__main__	Namespace(config='torchdistill/configs/legacy/sample/cifar10/ce/resnet20-final_run.yaml', device='cuda', log='log/cifar10/ce/resnet20-final_run.log', start_epoch=0, seed=None, test_only=False, student_only=False, log_config=False, world_size=1, dist_url='env://', adjust_lr=False)
2023/08/03 15:38:45	INFO	torchdistill.datasets.util	Loading train data
Downloading https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz to ./resource/dataset/cifar10/cifar-10-python.tar.gz
100% 170498071/170498071 [00:10<00:00, 16743363.48it/s]
Extracting ./resource/dataset/cifar10/cifar-10-python.tar.gz to ./resource/dataset/cifar10
2023/08/03 15:38:58	INFO	torchdistill.datasets.util	dataset_id `cifar10/train`: 13.498707056045532 sec
2023/08/03 15:38:58	INFO	torchdistill.datasets.util	Loading val data
Files already downloaded and verified
2023/08/03 15:38:59	INFO	torchdistill.datasets.util	dataset_id `cifar10/val`: 0.7275292873382568 sec
2023/08/03 15:38:59	INFO	torchdistill.datasets.util	Loading test data
Files already downloaded and verified
2023/08/03 15:39:00	INFO	torchdistill.datasets.util	dataset_id `cifar10/test`: 0.6866748332977295 sec
2023/08/03 15:39:00	INFO	torchdistill.common.main_util	ckpt file is not found at `./resource/ckpt/cifar10/ce/cifar10-resnet20-final_run.pt`
2023/08/03 15:39:02	INFO	__main__	Start training
/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:563: UserWarning: This DataLoader will create 16 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(
2023/08/03 15:39:02	INFO	torchdistill.models.util	[student model]
2023/08/03 15:39:02	INFO	torchdistill.models.util	Using the original student model
2023/08/03 15:39:02	INFO	torchdistill.core.training	Loss = 1.0 * OrgLoss
2023/08/03 15:39:05	INFO	torchdistill.misc.log	Epoch: [0]  [  0/391]  eta: 0:15:33  lr: 0.1  img/s: 115.78139357541274  loss: 2.6344 (2.6344)  time: 2.3876  data: 1.2818  max mem: 281
2023/08/03 15:39:10	INFO	torchdistill.misc.log	Epoch: [0]  [100/391]  eta: 0:00:22  lr: 0.1  img/s: 2267.8138509303653  loss: 1.7821 (1.9676)  time: 0.0710  data: 0.0058  max mem: 281
2023/08/03 15:39:17	INFO	torchdistill.misc.log	Epoch: [0]  [200/391]  eta: 0:00:14  lr: 0.1  img/s: 3803.279342589969  loss: 1.6063 (1.8255)  time: 0.0547  data: 0.0080  max mem: 281
2023/08/03 15:39:22	INFO	torchdistill.misc.log	Epoch: [0]  [300/391]  eta: 0:00:05  lr: 0.1  img/s: 3112.6740762643567  loss: 1.4312 (1.7326)  time: 0.0511  data: 0.0057  max mem: 281
2023/08/03 15:39:26	INFO	torchdistill.misc.log	Epoch: [0] Total time: 0:00:23
2023/08/03 15:39:28	INFO	torchdistill.misc.log	Validation:  [ 0/79]  eta: 0:01:49  acc1: 59.3750 (59.3750)  acc5: 93.7500 (93.7500)  time: 1.3893  data: 1.3608  max mem: 281
2023/08/03 15:39:31	INFO	torchdistill.misc.log	Validation: Total time: 0:00:05
2023/08/03 15:39:31	INFO	__main__	 * Acc@1 49.5600	Acc@5 94.1200

2023/08/03 15:39:31	INFO	__main__	Best top-1 accuracy: 0.0000 -> 49.5600
2023/08/03 15:39:31	INFO	__main__	Updating ckpt at ./resource/ckpt/cifar10/ce/cifar10-resnet20-final_run.pt
2023/08/03 15:39:33	INFO	torchdistill.misc.log	Epoch: [1]  [  0/391]  eta: 0:10:23  lr: 0.1  img/s: 943.0203439251023  loss: 1.2175 (1.2175)  time: 1.5934  data: 1.4574  max mem: 281
2023/08/03 15:39:39	INFO	torchdistill.misc.log	Epoch: [1]  [100/391]  eta: 0:00:21  lr: 0.1  img/s: 2045.4562883377148  loss: 1.2161 (1.2695)  time: 0.0517  data: 0.0050  max mem: 281
2023/08/03 15:39:44	INFO	torchdistill.misc.log	Epoch: [1]  [200/391]  eta: 0:00:12  lr: 0.1  img/s: 2955.1659694394293  loss: 1.1407 (1.2134)  time: 0.0759  data: 0.0053  max mem: 281
2023/08/03 15:39:52	INFO	torchdistill.misc.log	Epoch: [1]  [300/391]  eta: 0:00:06  lr: 0.1  img/s: 2962.160811727901  loss: 1.0447 (1.1683)  time: 0.0503  data: 0.0063  max mem: 281
2023/08/03 15:39:56	INFO	torchdistill.misc.log	Epoch: [1] Total time: 0:00:24
2023/08/03 15:39:57	INFO	torchdistill.misc.log	Validation:  [ 0/79]  eta: 0:01:28  acc1: 61.7188 (61.7188)  acc5: 99.2188 (99.2188)  time: 1.1196  data: 1.0868  max mem: 281
2023/08/03 15:39:59	INFO	torchdistill.misc.log	Validation: Total time: 0:00:03
2023/08/03 15:39:59	INFO	__main__	 * Acc@1 63.4100	Acc@5 96.4500

2023/08/03 15:39:59	INFO	__main__	Best top-1 accuracy: 49.5600 -> 63.4100
2023/08/03 15:39:59	INFO	__main__	Updating ckpt at ./resource/ckpt/cifar10/ce/cifar10-resnet20-final_run.pt
2023/08/03 15:40:02	INFO	torchdistill.misc.log	Epoch: [2]  [  0/391]  eta: 0:17:13  lr: 0.1  img/s: 959.3044746143098  loss: 0.9188 (0.9188)  time: 2.6443  data: 2.5106  max mem: 281
```

> I have made some modifications according to your request, and I think this should be correct. about [this script](https://github.com/lansiyuan3/Modulation/blob/main/torchdistill-main/torchdistill-main/examples/legacy/image_classification.py) and [yaml](https://github.com/lansiyuan3/Modulation/blob/main/torchdistill-main/torchdistill-main/configs/legacy/sample/cifar10/kd/resnet20_from_densenet_bc_k12_depth100-final_run.yaml)
I just want to experiment with whether this system can run and then add my own data loader, but I have found the following issues

Why don't you describe what modifications you made exactly? You just keep saying you made modifications, but never explain what modifications you made and how or differences from the legacy script in this repo either. You didn't even confirm that you used torchdistill=v0.3.3. I am not your personal tutor, and If you refuse to do those, I can not answer your questions anymore.

**Reply by lansiyuan3** [reply]
Sorry to bother you, but I think I have figured out what the problem is and sincerely apologize for my low-level mistake
